File: 03_week/02_acync_handling.clj
(ns first.core)

(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))


(def post {:id 66, :title "함수형 자바스크립트 스터디", :price 10000,
             :items [
                     {:no 0, :type "text", :body "커링과 파이프라인 코딩", :id nil},
                     {:no 1, :type "photo", :url "/img/curry.png", :id 11},
                     {:no 2, :type "heading", :body "함수형으로의 전환", :id nil},
                     {:no 3, :type "photo", :url "/img/profile.jpg", :id 23}]})


(defn db_api_update [x y]
  (Thread/sleep 1000)
  (println (str "db update: " x " where : " y))
  x)

(defn db_api_insert [x]
  (Thread/sleep 2000)
  (println (str "db insert: " x ))
  x)


(->> (:items post)
     (reduce
       (fn [res item]
         (if (nil? (:id item))
           (merge-with into res {:id_nil [(dissoc item :id)]})
           (merge-with into res {:has_id [item]})))
       {:has_id [] :id_nil []})
     ((fn [items]
        (let [has_id (:has_id items) id_nil (:id_nil items)]
          (-> (pcalls
                #(doall (pmap (fn [item]
                                (db_api_update (dissoc item :id) (:id item))) has_id))
                #(db_api_insert id_nil))
              doall
              ((partial println "end : ")))))))

File: 03_week/02_acync_handling.js
var post = {
  id: 66,
  title: '함수형 자바스크립트 스터디',
  price: 10000,
  _: { items: [
    { no: 0,
      type: 'text',
      body: '커링과 파이프라인 코딩',
      id: null },
    { no: 1,
      type: 'photo',
      url: '/img/curry.png',
      id: 11 },
    { no: 2,
      type: 'heading',
      body: '함수형으로의 전환',
      id: null },
    { no: 3,
      type: 'photo',
      url: '/img/profile.jpg',
      id: 23 }]
  }
};

var db_api_insert = function(table, set) {
  return new Promise(function(resolve){
    setTimeout(function(){
      console.log("db insert: ", set);
      resolve(set)
    }, 2000)
  })
};

var db_api_update = function(table, set, where) {
  return new Promise(function(resolve){
    setTimeout(function(){
      console.log("db update: ", set, where);
      resolve(set)
    }, 3000)
  })
};

_.go(post,
  post => post._.items,
  _.reduce(function(res, li){
    if (li.id) return res.has_id.push(li), res;
    return res.id_null.push(_.omit(li, 'id')), res;
  }, { has_id : [], id_null : [] }),
  _.all(
    __(
      items => items.has_id,
      _.map(item => db_api_update('item', _.omit(item, 'id'), { id: item.id }))),
    __(
      items => items.id_null,
      items => db_api_insert('item', items))),
  _(console.log, 'all end : '));


// _.go(post,
//   post => post._.items,
//   _.partition(_.val('id')),
//   _.to_mr,
//   _.spread(
//     _.map(item => db_api_update('item', _.omit(item, 'id'), { id: item.id })),
//     items=> db_api_insert('item', items)),
//   _(console.log, 'all end : '));